crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs

```diff
@@ -0,0 +1,202 @@
+        errors.push(unreferenced_binding);
+    };
+
+    // case - bool operations (`if test1 and test2:`)
+    if let Expr::BoolOp(expr) = if_test {
+        errors.extend(
+            expr.values
+                .iter()
+                .filter_map(|bool_test| {
+                    find_assignment_before_if_stmt(semantic, if_test, bool_test)
+                })
+                .collect::<Vec<AssignmentBeforeIfStmt>>(),
+        );
+    }
+
+    // case - compare (`if test1 is not None:`)
+    if let Expr::Compare(compare) = if_test {
+        if let Some(error) = find_assignment_before_if_stmt(semantic, if_test, &compare.left) {
+            errors.push(error);
+        };
+    }
+
+    // case - elif else clauses (`elif test1:`)
+    errors.extend(
+        stmt.elif_else_clauses
+            .iter()
+            .filter(|elif_else_clause| elif_else_clause.test.is_some())
+            .filter_map(|elif_else_clause| {
+                let elif_check = elif_else_clause.test.as_ref().unwrap();
+                find_assignment_before_if_stmt(semantic, elif_check, elif_check)
+            })
+            .collect::<Vec<AssignmentBeforeIfStmt>>(),
+    );
+
+    // add found diagnostics
+    checker.diagnostics.extend(
+        errors
+            .into_iter()
+            .map(|error| {
+                create_diagnostic(
+                    checker.locator(),
+                    checker.indexer(),
+                    checker.generator(),
+                    error,
+                )
+            })
+            .collect::<Vec<Diagnostic>>(),
+    );
+}
+
+/// Find possible assignment before if statement
+///
+/// * `if_test` - the complete if test (`if test1 and test2`)
+/// * `if_test_part` - part of the if test (`test1`)
+fn find_assignment_before_if_stmt<'a>(
+    semantic: &'a SemanticModel,
+    if_test: &'a Expr,
+    if_test_part: &'a Expr,
+) -> Option<AssignmentBeforeIfStmt<'a>> {
+    // early exit when the test part is not a variable
+    let Expr::Name(test_variable) = if_test_part else {
+        return None;
+    };
+
+    let current_statement = semantic.current_statement();
+    let previous_statement = semantic
+        .previous_statement(current_statement)?
+        .as_assign_stmt()?;
+
+    // only care about single assignment target like `x = 'example'`
+    let [assigned_variable] = &previous_statement.targets[..] else {
+        return None;
+    };
+
+    // check whether the check variable is the assignment variable
+    if test_variable.id != assigned_variable.as_name_expr()?.id {
+        return None;
+    }
+
+    Some((if_test, test_variable, previous_statement))
+}
+
+fn create_diagnostic(
+    locator: &Locator,
+    indexer: &Indexer,
+    generator: Generator,
+    error: AssignmentBeforeIfStmt,
+) -> Diagnostic {
+    let (origin, expr_name, assignment) = error;
+    let assignment_expr = generator.expr(&assignment.value);
```

Comment:
What's the motivation for calling into the generator here? We try to avoid using the generator because it removes comments. Could we instead take the assignment value as it is in the source (using locator)? Note: We have to be careful about parenthesized expressions.